docs/3.x/extend/graphql.md

```diff
@@ -0,0 +1,1169 @@
+
+```php
+use craft\events\RegisterGqlSchemaComponentsEvent;
+use craft\services\Gql;
+use yii\base\Event;
+
+Event::on(
+    Gql::class,
+    Gql::EVENT_REGISTER_GQL_SCHEMA_COMPONENTS,
+    function(RegisterGqlSchemaComponentsEvent $event) {
+        $event->queries = array_merge($event->queries, [
+            // “Widgets” group
+            'Widgets' => [
+                // widget component with read action, labelled “View Widgets” in UI
+                'widget:read' => ['label' => 'View Widgets']
+            ],
+        ]);
+
+        // Same format applies for $event->mutations
+    }
+);
+```
+
+If you’d like to further specify individual types, you can append each relevant UID in the key:
+
+```php
+use mynamespace\Plugin;
+use craft\events\RegisterGqlSchemaComponentsEvent;
+use craft\services\Gql;
+use yii\base\Event;
+
+Event::on(
+    Gql::class,
+    Gql::EVENT_REGISTER_GQL_SCHEMA_COMPONENTS,
+    function(RegisterGqlSchemaComponentsEvent $event) {
+        // Assume an array of objects each having `uid` and `name` properties
+        $widgetTypes = Plugin::getInstance()->getAllWidgetTypes();
+
+        if (!empty($widgetTypes)) {
+            $queryComponents = [];
+
+            foreach ($widgetTypes as $widgetType) {
+                $queryComponents['widget.' . $widgetType->uid . ':read'] = [
+                    'label' => 'View Widgets – ' . $widgetType->name
+                ];
+            }
+        }
+
+        $event->queries = array_merge($event->queries, [
+            // “Widgets” group
+            'Widgets' => $queryComponents,
+        ]);
+    }
+);
+```
+
+### Eager Loading
+
+If you’re adding relational fields that would benefit from eager loading, you’ll want to register those by adding them to the `fieldList` array on the [registerGqlEagerLoadableFields](craft3:craft\gql\ElementQueryConditionBuilder::EVENT_REGISTER_GQL_EAGERLOADABLE_FIELDS) event object.
+
+::: tip
+Any fields you add using Craft’s stock GraphQL types do _not_ need to be registered for eager loading.
+:::
+
+If we had added a custom field with the handle `widgets`, we’d register it like this:
+
+```php
+use mynamespace\fields\Widgets;
+use craft\gql\ElementQueryConditionBuilder;
+use craft\events\RegisterGqlEagerLoadableFields;
+use yii\base\Event;
+
+Event::on(
+    ElementQueryConditionBuilder::class,
+    ElementQueryConditionBuilder::EVENT_REGISTER_GQL_EAGERLOADABLE_FIELDS,
+    function(RegisterGqlEagerLoadableFields $event) {
+        $event->fieldList['widgets'] = [Widgets::class];
+    }
+);
+```
+
+The key should be the field name, and the value can be one or more classes that define it.
+
+Any classes you provide determine the context for that field to be eager-loadable. You may otherwise pass `*` which, if present, allows that field to be used and eager loaded anywhere at all:
+
+```php
+// ...
+$event->fieldList['widgets'] = ['*'];
+// ...
+```
+
+You can use the `canBeAliased` key to do something:
```

Comment:
Probably worth rewording.

Reply:
Sorry, I meant to remove this and the example—and somehow managed not to. 🤦